Adjust `BeatSyncContainer`'s early animate offset based on source's rate

peppy opened the pull request:
As discussed in https://github.com/ppy/osu/discussions/26644.

I tried to add test coverage for this but it will need a completely new setup (and real time considerations) to check this in real-world conditions. Probably fine to manually test this one?

attempt:

```diff
diff --git a/osu.Game.Tests/Visual/UserInterface/TestSceneBeatSyncedContainer.cs b/osu.Game.Tests/Visual/UserInterface/TestSceneBeatSyncedContainer.cs
index c723610614..f026c880f3 100644
--- a/osu.Game.Tests/Visual/UserInterface/TestSceneBeatSyncedContainer.cs
+++ b/osu.Game.Tests/Visual/UserInterface/TestSceneBeatSyncedContainer.cs
@@ -13,6 +13,7 @@
 using osu.Framework.Graphics.Containers;
 using osu.Framework.Graphics.Shapes;
 using osu.Framework.Testing;
+using osu.Framework.Timing;
 using osu.Framework.Utils;
 using osu.Game.Beatmaps.ControlPoints;
 using osu.Game.Graphics;
@@ -170,6 +171,41 @@ public void TestEarlyActivationEffectPoint(bool earlyActivating)
             AddAssert("effect has kiai", () => actualEffectPoint != null && ((EffectControlPoint)actualEffectPoint).KiaiMode);
         }
 
+        [TestCase(true)]
+        [TestCase(false)]
+        public void TestEarlyActivationWithRate(bool drawableClockHasRate)
+        {
+            const double early_activation_milliseconds = 100;
+
+            ControlPoint actualEffectPoint = null;
+
+            AddStep($"set early activation to {early_activation_milliseconds}", () => beatContainer.EarlyActivationMilliseconds = early_activation_milliseconds);
+
+            if (!drawableClockHasRate)
+                AddStep("Hide rate from container", () => beatContainer.Clock = new NoRateClock(gameplayClockContainer.Clock));
+
+            AddStep("Set rate extreme", () => gameplayClockContainer.UserPlaybackRate.Value = 0.4f);
+
+            AddStep("seek before kiai effect point", () =>
+            {
+                ControlPoint expectedEffectPoint = Beatmap.Value.Beatmap.ControlPointInfo.EffectPoints.First(ep => ep.KiaiMode);
+                actualEffectPoint = null;
+                beatContainer.AllowMistimedEventFiring = false;
+
+                beatContainer.NewBeat = (_, _, effectControlPoint, _) =>
+                {
+                    if (Precision.AlmostEquals(gameplayClockContainer.CurrentTime + early_activation_milliseconds * gameplayClockContainer.Rate, expectedEffectPoint.Time, BeatSyncedContainer.MISTIMED_ALLOWANCE))
+                        actualEffectPoint = effectControlPoint;
+                };
+
+                gameplayClockContainer.Seek(expectedEffectPoint.Time - early_activation_milliseconds);
+            });
+
+            AddUntilStep("wait for effect point", () => actualEffectPoint != null);
+
+            AddAssert("effect has kiai", () => actualEffectPoint != null && ((EffectControlPoint)actualEffectPoint).KiaiMode);
+        }
+
         private partial class TestBeatSyncedContainer : BeatSyncedContainer
         {
             private const int flash_layer_height = 150;
@@ -338,4 +374,29 @@ public InfoString(string header)
             }
         }
     }
+
+    public class NoRateClock : IFrameBasedClock
+    {
+        private readonly IFrameBasedClock clock;
+
+        public NoRateClock(IFrameBasedClock clock)
+        {
+            this.clock = clock;
+        }
+
+        public double CurrentTime => clock.CurrentTime;
+
+        public double Rate => 1;
+
+        public bool IsRunning => clock.IsRunning;
+
+        public void ProcessFrame()
+        {
+            clock.ProcessFrame();
+        }
+
+        public double ElapsedFrameTime => clock.ElapsedFrameTime;
+
+        public double FramesPerSecond => clock.FramesPerSecond;
+    }
 }

```

peppy committed:
Adjust `BeatSyncContainer`'s early animate offset based on source's rate

bdach commented on osu.Game/Graphics/Containers/BeatSyncedContainer.cs:
```diff
@@ -91,7 +91,17 @@ protected override void Update()
+                // But other cases like the osu! logo at the main menu won't correctly have this rate information.
+                //
+                // So for cases where the rate of the source isn't in sync with our hierarchy, let's assume we need to account for it locally.
+                if (Clock.Rate == 1 && BeatSyncSource.Clock.Rate != Clock.Rate)
```
The general idea here appears to check out but I am left wondering about the `Clock.Rate == 1` check here. Is this not an issue for any clock rate mismatch? And as such the early activation period should be adjusted by a ratio of `BeatSyncSource.Clock.Rate / Clock.Rate`?

peppy replied:
I thought I replied to this way back but it must have been in my dreams, or lost as a draft..

Re-reading the comment above, I think what I have makes sense. It's only aiming to apply a correction when it's very clear that we haven't already taken on a hierarchical rate adjustment (because if we have, it's assumed that things are already in a good state).

Your proposal may work though, since in the case we *have* got a hierarchical rate, it would be `early *= 1`.

I think your proposal is to change this to:

```diff
diff --git a/osu.Game/Graphics/Containers/BeatSyncedContainer.cs b/osu.Game/Graphics/Containers/BeatSyncedContainer.cs
index a14dfd4d64..de9a5aff3a 100644
--- a/osu.Game/Graphics/Containers/BeatSyncedContainer.cs
+++ b/osu.Game/Graphics/Containers/BeatSyncedContainer.cs
@@ -96,10 +96,8 @@ protected override void Update()
                 // In the case of gameplay, we are usually within a hierarchy with the correct rate applied to our `Drawable.Clock`.
                 // This means that the amount of early adjustment is adjusted in line with audio track rate changes.
                 // But other cases like the osu! logo at the main menu won't correctly have this rate information.
-                //
-                // So for cases where the rate of the source isn't in sync with our hierarchy, let's assume we need to account for it locally.
-                if (Clock.Rate == 1 && BeatSyncSource.Clock.Rate != Clock.Rate)
-                    early *= BeatSyncSource.Clock.Rate;
+                // We can adjust here to ensure the applied early activation always matches expectations.
+                early *= BeatSyncSource.Clock.Rate / Clock.Rate;
 
                 currentTrackTime = BeatSyncSource.Clock.CurrentTime + early;
 

```

?

bdach replied:
That would be the proposal I think yeah.

peppy replied:
Seems to work well enough, have applied.

bdach replied:
https://github.com/ppy/osu/pull/26707/commits/e02ad6cf94d3a9bba0cbcac6ecee5a0125f9da89 has me thinking that maybe a guard against zero rate is warranted somewhere closer to this change rather than doing test changes...

peppy replied:
dunno. i don't think we should ever be hitting zero rate anymore. your call though.

bdach replied:
I've pushed a guard (https://github.com/ppy/osu/pull/26707/commits/0b29a762b8aa712d1a0b697f367ae809a0896f15) because I have the unshakable feeling that if we don't now we're gonna have to hotfix this later.

peppy committed:
Merge branch 'master' into beat-sync-container-early-adjust

peppy committed:
Apply rate adjust fix in all cases rather than specifically for `Clock.Rate == 1`

peppy committed:
Fix test regression

bdach committed:
Add precautionary guard to avoid potential div-by-zero

Probably wouldn't happen outside of tests, but I'd rather not find out
next release.
